float64 x_ref
float64[] uff
float64[] k